cpp/src/arrow/filesystem/azurefs.cc

```diff
@@ -317,27 +321,136 @@ class ObjectInputFile final : public io::RandomAccessFile {
 class AzureFileSystem::Impl {
  public:
   io::IOContext io_context_;
-  std::shared_ptr<Azure::Storage::Blobs::BlobServiceClient> service_client_;
+  std::shared_ptr<Azure::Storage::Files::DataLake::DataLakeServiceClient>
+      datalake_service_client_;
+  std::shared_ptr<Azure::Storage::Blobs::BlobServiceClient> blob_service_client_;
```

Comment:
👍 If I understand correctly the S3 and GCS filesystems it had to use shared pointers because the top level clients get passed to the input file and output stream classes but for Azure we only pass the blob client to those classes so `blob_service_client_` can beunique. `datalake_service_client_` though is shared with the `HierarchicalNamespaceDetector` so needs to be shared.